Lib/test/test_os.py

```diff
@@ -4485,6 +4485,57 @@ def test_openpty(self):
         self.assertEqual(os.get_inheritable(master_fd), False)
         self.assertEqual(os.get_inheritable(slave_fd), False)
 
+    @unittest.skipUnless(hasattr(os, 'spawnl'), "need os.openpty()")
+    def test_pipe_spawnl(self):
+        # gh-77046: On Windows, os.pipe() file descriptors must be created with
+        # _O_NOINHERIT to make them non-inheritable. UCRT has no public API to
+        # get (_osfile(fd) & _O_NOINHERIT), so use a functional test.
+        fd, fd2 = os.pipe()
+        self.addCleanup(os.close, fd)
+        self.addCleanup(os.close, fd2)
+
+        # Make sure that fd is not inherited by a child process created by
+        # os.spawnl(): get_osfhandle() and dup() must fail with EBADF.
+        code = textwrap.dedent(f"""
+            import errno
+            import os
+            import test.support
+            try:
+                import msvcrt
+            except ImportError:
+                msvcrt = None
+
+            fd = {fd}
+
+            with test.support.SuppressCrashReport():
+                if msvcrt is not None:
+                    try:
+                        handle = msvcrt.get_osfhandle(fd)
+                    except OSError as exc:
+                        if exc.errno != errno.EBADF:
+                            raise
+                    else:
+                        raise Exception("get_osfhandle() must fail")
+
```

Comment:
With `else` it would be clearer that the following code is not a fallback, but an alternative version.

Oh, and because almost all code is different, you can use different scripts, depending on the OS.

Reply:
> With else it would be clearer that the following code is not a fallback

Sorry, else on which line? Would you mind to elaborate?

The code calls msvcrt.get_osfhandle() **and** os.dup() on Windows on purpose. I explained it in a comment a few lines before.

> Oh, and because almost all code is different, you can use different scripts, depending on the OS.

I prefer to have a single code base on all platforms, since most of the code is in common. Only the code using `msvcrt` is optional. Even on Windows, `msvcrt` may not be available on Python implementations other than CPython.

Reply:
Ah, I see now that there is a possibility for the code above to not raise an exception. So the code below this line can be executed on Windows.

Reply:
The code makes sure that  msvcrt.get_osfhandle() and os.dup() both raise an exception on the "not inherited" file descriptor.